File: app/src/main/java/com/mytlogos/enterprise/ui/MediaInWaitListFragment.java
package com.mytlogos.enterprise.ui;

import android.annotation.SuppressLint;
import android.os.AsyncTask;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.ViewModelProviders;
import androidx.paging.PagedList;

import com.mytlogos.enterprise.R;
import com.mytlogos.enterprise.model.MediumInWait;
import com.mytlogos.enterprise.model.MediumType;
import com.mytlogos.enterprise.tools.Sortings;
import com.mytlogos.enterprise.tools.Utils;
import com.mytlogos.enterprise.viewmodel.MediaInWaitListViewModel;

import java.io.IOException;
import java.io.Serializable;
import java.util.ArrayList;
import java.util.Collection;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Objects;

import eu.davidea.flexibleadapter.FlexibleAdapter;
import eu.davidea.flexibleadapter.items.AbstractFlexibleItem;
import eu.davidea.flexibleadapter.items.IFlexible;

public class MediaInWaitListFragment extends BaseSwipeListFragment<MediumInWait, MediaInWaitListViewModel> {

    @NonNull
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = super.onCreateView(inflater, container, savedInstanceState);
        this.setTitle("Unused Media");
        return view;
    }

    @Nullable
    @Override
    Filterable createFilterable() {
        return new Filterable() {

            @Override
            public void onCreateFilter(View view, AlertDialog.Builder builder) {
                setMediumCheckbox(view, R.id.text_medium, MediumType.TEXT);
                setMediumCheckbox(view, R.id.audio_medium, MediumType.AUDIO);
                setMediumCheckbox(view, R.id.video_medium, MediumType.VIDEO);
                setMediumCheckbox(view, R.id.image_medium, MediumType.IMAGE);
            }

            @Override
            public int getFilterLayout() {
                return R.layout.filter_medium_in_wait_layout;
            }

            @Override
            public FilterProperty[] getSearchFilterProperties() {
                return new FilterProperty[]{
                        new FilterProperty() {
                            @Override
                            public int getSearchViewId() {
                                return R.id.title_filter;
                            }

                            @Override
                            public int getClearSearchButtonId() {
                                return R.id.clear_title;
                            }

                            @Override
                            public String get() {
                                return getViewModel().getTitleFilter();
                            }

                            @Override
                            public void set(String newFilter) {
                                getViewModel().setTitleFilter(newFilter);
                            }
                        },
                        new FilterProperty() {
                            @Override
                            public int getSearchViewId() {
                                return R.id.host_filter;
                            }

                            @Override
                            public int getClearSearchButtonId() {
                                return R.id.clear_host;
                            }

                            @Override
                            public String get() {
                                return getViewModel().getHostFilter();
                            }

                            @Override
                            public void set(String newFilter) {
                                getViewModel().setHostFilter(newFilter);
                            }
                        }
                };
            }

        };
    }

    @Override
    LinkedHashMap<String, Sortings> getSortMap() {
        LinkedHashMap<String, Sortings> map = new LinkedHashMap<>();
        map.put("Title A-Z", Sortings.TITLE_AZ);
        map.put("Title Z-A", Sortings.TITLE_ZA);
        map.put("Medium Asc", Sortings.MEDIUM);
        map.put("Medium Desc", Sortings.MEDIUM_REVERSE);
        map.put("Host A-Z", Sortings.HOST_AZ);
        map.put("Host Z-A", Sortings.HOST_ZA);
        return map;
    }

    @Override
    MediaInWaitListViewModel createViewModel() {
        return ViewModelProviders.of(this).get(MediaInWaitListViewModel.class);
    }

    @Override
    LiveData<PagedList<MediumInWait>> createPagedListLiveData() {
        return this.getViewModel().getMediaInWait();
    }

    @Override
    List<IFlexible> convertToFlexibles(Collection<MediumInWait> mediumItems) {
        List<IFlexible> items = new ArrayList<>();

        for (MediumInWait item : mediumItems) {
            if (item == null) {
                break;
            }
            items.add(new MediumItem(item));
        }
        return items;
    }

    @Override
    void onSwipeRefresh() {
        new LoadingTask().execute();
    }

    @Override
    public boolean onItemClick(View view, int position) {
        MediumItem item = (MediumItem) getFlexibleAdapter().getItem(position);

        if (item == null) {
            return false;
        }
        getMainActivity().switchWindow(MediaInWaitFragment.getInstance(item.item));
        return true;
    }

    @SuppressLint("StaticFieldLeak")
    private class LoadingTask extends AsyncTask<Void, Void, Void> {
        private String errorMsg = null;

        @Override
        protected Void doInBackground(Void... voids) {
            try {
                getViewModel().loadMediaInWait();
            } catch (IOException e) {
                errorMsg = "Loading went wrong";
                e.printStackTrace();
            }
            return null;
        }

        @Override
        protected void onPostExecute(Void aVoid) {
            if (this.errorMsg != null) {
                Toast.makeText(getContext(), errorMsg, Toast.LENGTH_SHORT).show();
            }
            getListContainer().setRefreshing(false);
        }
    }

    private static class MediumFilter implements Serializable {
        private final String title;
        private final int medium;
        private final String link;

        private MediumFilter(String title, int medium, String link) {
            this.title = title == null ? "" : title.toLowerCase();
            this.medium = medium;
            this.link = link == null ? "" : link.toLowerCase();
        }
    }

    private static class MediumItem extends AbstractFlexibleItem<MetaViewHolder> /*implements IFilterable<MediumFilter>*/ {
        private final MediumInWait item;

        MediumItem(@NonNull MediumInWait item) {
            this.item = item;
            this.setDraggable(false);
            this.setSwipeable(false);
            this.setSelectable(false);
        }

        @Override
        public int getLayoutRes() {
            return R.layout.meta_item;
        }

        @Override
        public MetaViewHolder createViewHolder(View view, FlexibleAdapter<IFlexible> adapter) {
            return new MetaViewHolder(view, adapter);
        }

        @SuppressLint("DefaultLocale")
        @Override
        public void bindViewHolder(FlexibleAdapter<IFlexible> adapter, MetaViewHolder holder, int position, List<Object> payloads) {
            String mediumType;

            switch (this.item.getMedium()) {
                case MediumType.AUDIO:
                    mediumType = "Audio";
                    break;
                case MediumType.IMAGE:
                    mediumType = "Bild";
                    break;
                case MediumType.TEXT:
                    mediumType = "Text";
                    break;
                case MediumType.VIDEO:
                    mediumType = "Video";
                    break;
                default:
                    String msg = String.format("no valid medium type: %d", this.item.getMedium());
                    throw new IllegalStateException(msg);
            }
            holder.topLeftText.setText(mediumType);

            String domain = Utils.getDomain(this.item.getLink());
            holder.topRightText.setText(domain);
            holder.mainText.setText(this.item.getTitle());
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;

            MediumItem that = (MediumItem) o;

            return Objects.equals(item, that.item);
        }

        @Override
        public int hashCode() {
            return item != null ? item.hashCode() : 0;
        }

        public boolean filter(MediumFilter constraint) {
            if (constraint == null) {
                return true;
            }
            if (constraint.medium > 0 && (this.item.getMedium() & constraint.medium) == 0) {
                return false;
            }
            if (!constraint.link.isEmpty() && !(this.item.getLink().toLowerCase().contains(constraint.link))) {
                return false;
            }
            return constraint.title.isEmpty() || this.item.getTitle().toLowerCase().contains(constraint.title);
        }
    }

}

File: app/src/main/java/com/mytlogos/enterprise/ui/MediaInWaitListFragment.kt
package com.mytlogos.enterprise.ui

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AlertDialog
import androidx.lifecycle.lifecycleScope
import androidx.paging.PagingData
import androidx.recyclerview.widget.DiffUtil
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.mytlogos.enterprise.R
import com.mytlogos.enterprise.model.*
import com.mytlogos.enterprise.tools.Sortings
import com.mytlogos.enterprise.tools.getDomain
import com.mytlogos.enterprise.viewmodel.MediaInWaitListViewModel
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.launch
import java.io.IOException
import java.util.*

class MediaInWaitListFragment : BaseSwipePagingFragment<MediumInWait, MediaInWaitListViewModel>() {
    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View {
        val view = super.onCreateView(inflater, container, savedInstanceState)
        this.setTitle("Unused Media")
        return view
    }

    override fun createFilterable(): Filterable {
        return object : Filterable {
            override val filterLayout: Int
                get() = R.layout.filter_medium_in_wait_layout

            override val searchFilterProperties: Array<Property<*>>
                get() = arrayOf(
                    object : TextProperty {
                        override val viewId: Int
                            get() = R.id.title_filter
                        override val clearViewId: Int
                            get() = R.id.clear_title

                        override fun get(): String {
                            return viewModel.titleFilter
                        }

                        override fun set(newFilter: String) {
                            viewModel.titleFilter = newFilter
                        }
                    },
                    object : TextProperty {
                        override val viewId: Int
                            get() = R.id.host_filter
                        override val clearViewId: Int
                            get() = R.id.clear_host

                        override fun get(): String {
                            return viewModel.hostFilter
                        }

                        override fun set(newFilter: String) {
                            viewModel.hostFilter = newFilter
                        }
                    }
                )
        }
    }

    override val sortMap: LinkedHashMap<String, Sortings>
        get() {
            val map = LinkedHashMap<String, Sortings>()
            map["Title A-Z"] = Sortings.TITLE_AZ
            map["Title Z-A"] = Sortings.TITLE_ZA
            map["Medium Asc"] = Sortings.MEDIUM
            map["Medium Desc"] = Sortings.MEDIUM_REVERSE
            map["Host A-Z"] = Sortings.HOST_AZ
            map["Host Z-A"] = Sortings.HOST_ZA
            return map
        }
    override val viewModelClass: Class<MediaInWaitListViewModel>
        get() = MediaInWaitListViewModel::class.java

    override fun onSwipeRefresh() {
        lifecycleScope.launch {
            try {
                @Suppress("BlockingMethodInNonBlockingContext")
                viewModel.loadMediaInWait()
            } catch (e: IOException) {
                showToast("Loading went wrong")
                e.printStackTrace()
            }
            (listContainer as SwipeRefreshLayout).isRefreshing = false
        }
    }

    override fun onItemClick(position: Int, item: MediumInWait?) {
        if (item == null) {
            return
        }
        mainActivity.switchWindow(MediaInWaitFragment.getInstance(item))
        return
    }

    override fun createAdapter(): BaseAdapter<MediumInWait, *> {
        return MediumInWaitAdapter()
    }

    @ExperimentalCoroutinesApi
    override fun createPaged(model: MediaInWaitListViewModel): Flow<PagingData<MediumInWait>> {
        return model.mediaInWait
    }

    class MediumInWaitHelper : DiffUtil.ItemCallback<MediumInWait>() {
        override fun areItemsTheSame(oldItem: MediumInWait, newItem: MediumInWait): Boolean {
            return oldItem == newItem
        }

        override fun areContentsTheSame(oldItem: MediumInWait, newItem: MediumInWait): Boolean {
            return oldItem == newItem
        }
    }

    private class MediumInWaitAdapter :
        BaseAdapter<MediumInWait, NewMetaViewHolder>(MediumInWaitHelper()) {

        override val layoutId: Int
            get() = R.layout.meta_item

        override fun createViewHolder(root: View, viewType: Int) = NewMetaViewHolder(root)

        override fun onBindViewHolder(holder: NewMetaViewHolder, position: Int) {
            getItem(position)?.let {
                val mediumType: String = when (it.medium) {
                    AUDIO -> "Audio"
                    IMAGE -> "Bild"
                    TEXT -> "Text"
                    VIDEO -> "Video"
                    else -> throw IllegalStateException("no valid medium type: ${it.medium}")
                }
                val domain = getDomain(it.link)

                holder.topLeftText.text = mediumType
                holder.topRightText.text = domain
                holder.mainText.text = it.title
            }
        }
    }
}